### ISSUE TITLE
various write tcp and read tcp errors when executing restic > rest-server

### ISSUE THREAD

**Issue by deajan** [restic/restic#4125]
Output of `restic version`
--------------------------
restic 0.14.0 compiled with go1.19 on linux/amd64

How did you run restic exactly?
-------------------------------

```
export RESTIC_PASSWORD=mypassword
export RESTIC_REPOSITORY=rest:https://user:pass@someserver.tld/user

/usr/local/bin/restic backup --compression=auto --files-from-verbatim "${FILE_LIST}" --tag "sometag" -o rest.connections=15 -v
/usr/local/bin/restic restore cb698a9d -t /restore -o rest.connections=20
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
rest-server version rest-server 0.11.0 compiled with go1.17.6 on linux/amd64

Expected behavior
-----------------

Normal logs

Actual behavior
---------------
I get a couple of write tcp and read tcp errors while doing backup / restore operations.

Backups
```
/usr/local/bin/restic backup --compression=auto --files-from-verbatim "${FILE_LIST}" --tag "sometag" -o rest.connections=15 -v

[...]
Save(<data/a4a6f93bec>) returned error, retrying after 612.750245ms: client.PutObject: Put https://somehost/data/48/[...long str...]: write tcp [2a01:4xxx:xxxx:xxxx::Y]:19081->[2a01:xxxx:yyyy:zzzz::bac]:443: use of closed network connection
[...]
```

Restores
```
/usr/local/bin/restic restore cb698a9d -t /restore -o rest.connections=20

[...]
repository b3ffdd61 opened (repository version 2) successfully, password is correct
restoring <Snapshot cb697a0d of [/tmp/machine01.warp.local.xml /tmp/machine02.warp.local.xml /tmp/machine03.warp.local.xml /tmp/machine04.warp.local.xml /tmp/machine05.local.xml /tmp/machine06.warp.local.xml /var/lib/libvirt/images/machine01.warp.local.qcow2 /var/lib/libvirt/images/machine02.local.qcow2 /data/qemu/disks/machine02-data.warp.local /var/lib/libvirt/images/machine03.warp.local.qcow2 /var/lib/libvirt/images/machine04.warp.local.qcow2 /var/lib/libvirt/images/machine05.warp.local.qcow2 /var/lib/libvirt/images/machine06.warp.local.qcow2 /tmp/CUBE-BACKUP.1665918840.manifest] at 2023-01-02 13:37:39.011284922 +0200 CEST by root@vth> to /data/restrest
ignoring error for /var/lib/libvirt/images/meshcentral.warp.local.qcow2: StreamPack: ReadFull: read tcp [2a01:4xxx:xxxx:xxxx::Y]:38954->[2a01:xxxx:yyyy:zzzz::bac]:443: read: connection reset by peer
ignoring error for /var/lib/libvirt/images/opnsense.warp.local.qcow2: StreamPack: ReadFull: read tcp [2a01:4xxx:xxxx:xxxx::Y]:38954->[2a01:xxxx:yyyy:zzzz::bac]:443: read: connection reset by peer
ignoring error for /var/lib/libvirt/images/smailer.warp.local.qcow2: StreamPack: ReadFull: read tcp [2a01:4xxx:xxxx:xxxx::Y]:38954->[2a01:xxxx:yyyy:zzzz::bac]:443: read: connection reset by peer
ignoring error for /data/qemu/disks/cloud.sb-data.warp.local: StreamPack: ReadFull: read tcp [2a01:4xxx:xxxx:xxxx::Y]:38954->[2a01:xxxx:yyyy:zzzz::bac]:443: read: connection reset by peer
ignoring error for /var/lib/libvirt/images/cloud.sb.warp.local.qcow2: StreamPack: ReadFull: read tcp [2a01:4xxx:xxxx:xxxx::Y]:38954->[2a01:xxxx:yyyy:zzzz::bac]:443: read: connection reset by peer
ignoring error for /var/lib/libvirt/images/dmail.sb.warp.local.qcow2: StreamPack: ReadFull: read tcp [2a01:4xxx:xxxx:xxxx::Y]:38954->[2a01:xxxx:yyyy:zzzz::bac]:443: read: connection reset by peer
Fatal: There were 6 errors
[...]
```

Steps to reproduce the behavior
-------------------------------

Do you have any idea what may have caused this?
-----------------------------------------------

I run a haproxy in front of the restic server, which offloads TLS.

default config
```
defaults
    mode                    http
    log                     global
    option                  httplog
    option                  dontlognull
    option http-server-close
    option forwardfor       except 127.0.0.0/8
    option                  redispatch
    retries                 3
    timeout http-request    10s
    timeout queue           1m
    timeout connect         10s
    timeout client          1m
    timeout server          1m
    timeout http-keep-alive 10s
    timeout check           10s
    maxconn                 1500
```
backend config as follows:
```
backend rest_server
        option http-keep-alive
        option forwardfor
        server restic_storage                     10.15.XX.YY:ZZZZ
```

Do you have an idea how to solve the issue?
-------------------------------------------

I've read on the somewhere on the forum that those errors can be discarded since restic still manages to backup / restore, and only retries until read/writes are successful, or totally fails.

Since I wanted to trust these statements, I made sha256 sums of all backed up files, and checked that those correspond to the restored files. Checks out.

While I'd be happy to make those error messages more clear (like adding "retrying" at the end of the error), I'd also love to know what generates those errors.
Links between restic client and rest-server are about 300Mbps with iperf.

Can those problems come from `-o rest.connections=x` value ?
Maybe from haproxy ? Is there any previous record of a problem with haproxy ?
Maybe sh

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Getting to love restic ;)
I'm building both virsh backup and portable windows backup tools right now.

**Comment by MichaelEischer** [comment]
> Save(<data/a4a6f93bec>) returned error, retrying after 612.750245ms: client.PutObject: Put https://somehost/data/48/[...long str...]: write tcp [2a01:4xxx:xxxx:xxxx::Y]:19081->[2a01:xxxx:yyyy:zzzz::bac]:443: use of closed network connection

That looks like haproxy closes the connection from time to time. Maybe `http-server-close` has something to do with that? Or the `http-keep-alive` option?

> While I'd be happy to make those error messages more clear (like adding "retrying" at the end of the error),

That will be fixed in restic 0.15.0 (https://github.com/restic/restic/pull/2661)

**Comment by deajan** [comment]
@MichaelEischer Thank you.
I'll play around with haproxy and report back if I find something.

> That will be fixed in restic 0.15.0 (https://github.com/restic/restic/pull/2661)

Awesome :)

**Comment by MichaelEischer** [comment]
Closing as this seems to be a problem of the server setup. Feel free to open a new issue if you still think the problem is in restic.
